content/blog/2018/appswitch/index.md

```diff
@@ -0,0 +1,81 @@
+
+Dependency ordering is deceptively nuanced with a host of interrelated problems.  Ordering individual services requires having the dependency graph of the services so that they can be brought up starting from leaf nodes back to the root nodes.  It is not easy to construct such a graph and keep it updated over time as interdependencies evolve with the behavior of the application.  Even if the dependency graph is somehow provided, enforcing the ordering itself is not easy.  Simply starting the services in the specified order obviously won’t do.  A service may have started but not be ready to accept connections yet.  This is the problem with docker-compose's `depends-on` tag, for example.
+
+Apart from introducing sufficiently long sleeps between service startups, a common pattern that is often used is to check for readiness of dependencies before starting a service.  In Kubernetes, this is typically done as part of the init container of the pod.  However that means that the entire application would be held up until all its dependencies come alive.  Sometimes applications spend several minutes initializing themselves on startup before making their first outbound connection.  Not allowing a service to start adds substantial overhead to overall startup time of the application.  Also, checking for readiness of dependencies won’t work in case of "multilevel" interdependencies where a service needs to be started before another service but only after a third service.  Likewise, multiple containers of the same pod may have mutual dependencies that cannot be ordered by simply waiting at the init container.
+
+### Example scenario: IBM WebSphere ND
+
+Let us consider IBM WebSphere ND -- a widely deployed application middleware -- to grok these problems more closely.  It is a fairly complex framework in itself and consists of a central component called deployment manager (`dmgr`) that manages a set of node instances.  It uses UDP to negotiate cluster membership among the nodes and requires that deployment manager is up and operational before any of the node instances can come up and join the cluster.
+
+Why are we talking about a traditional application in the modern cloud-native context?  It turns out that there are significant gains to be had by enabling them to run on the Kubernetes and Istio platforms.  Essentially it's a part of the modernization journey that allows running traditional apps alongside green-field apps on the same modern platform to facilitate interoperation between the two.  As such, running WebSphere ND on Kubernetes requires support for projecting a consistent network environment with specific network interface attributes etc. to the application for it to come up and run correctly and AppSwitch is designed to take care of that.  For the purpose of this blog however, I'll focus on the application's dependency ordering requirement and how AppSwitch addresses it.
+
```

Comment:
Please describe how websphere is deployed (as stateful set or as headless service or as a regular k8s service), to ground the discussion

Reply:
>  As such, running WebSphere ND on Kubernetes requires support for projecting a consistent network environment with specific network interface attributes etc. to the application for it to come up and run correctly and AppSwitch is designed to take care of that.

this sentence stands out and could be killed (its rather long and a bit vague)

Reply:
Yes it turned out to be a bit of a run on.  Let me rework it.  I included it because dependency ordering is not the first thing that jumps out when talking about traditional apps on k8s.  There are several other things that get in the way which need to be taken care.